### PR TITLE
Implement includes for the mongo ext_pillar.

### PR DESCRIPTION
**Pull request by jochumdev**
### What does this PR do?

A WIP feature which adds the possibility to include files from other pillars for the mongo ext_pillar.

Please give feedback, i have the following questions:
- Is it usefull, something that we want in core?
- Is there a better way to handle from which environment we include future includes instead of 
  defining it per include.

This is an example of such a mongodb document that i use with includes:

```
{
    "_id" : "srv02.pcdummy.lan",
    "include" : [ 
        {
            "file" : "roles.base.server",
            "saltenv" : "pcdummy"
        }, 
        {
            "file" : "roles.base.lxc",
            "saltenv" : "pcdummy"
        }, 
        {
            "file" : "roles.base.postfix-relayclient",
            "saltenv" : "pcdummy"
        }
    ],
    "lxc" : {
        "default_conf" : [ 
            {
                "lxc:network:type" : "veth"
            }, 
            {
                "lxc:network:link" : "lanbr0"
            }, 
            {
                "lxc:network:flags" : "up"
            }, 
            {
                "lxc:network:hwaddr" : "00:16:3e:04:xx:xx"
            }
        ],
        "users" : {
            "lxd" : {
                "interfaces" : {
                    "lanbr0" : {
                        "type" : "veth",
                        "count" : 100
                    }
                }
            }
        }
    }
}
```
### Tests written?

No, there are no tests for the mongo ext_pillar.

Signed-off-by: Rene Jochum rene@jochums.at

### CODE REVIEW

**Comment by pcdummy** [comment]
I just have written a HOWTO where i use that: https://rene.jochums.at/howto-move-saltstack-tops-and-pillar-contents-to-mongodb

**Comment by cachedout** [comment]
I can see how this would be useful. I would like to have documentation included in this PR though, before we merge it.

cc: @thatch45 @terminalmage

**Comment by rallytime** [comment]
Go Go Jenkins!

File: salt/pillar/mongo.py [hunk 1]
```diff
@@ -151,6 +161,78 @@ def ext_pillar(minion_id,
             )
         else:
             log.debug('ext_pillar.mongo: found document, returning whole doc')
+
+        if 'include' in result:
+
+            if not isinstance(result['include'], list):
+                msg = ('MongoDB include in Document {0!r} '
```
Comment (terminalmage) [inline salt/pillar/mongo.py]: Please do not use the repr flag here, change this line to:

```python
msg = ('MongoDB include in document \'{0}\' '
```

Comment (pcdummy) [inline salt/pillar/mongo.py]: done.

**Comment by pcdummy** [comment]
docs added.

File: salt/pillar/mongo.py [hunk 2]
```diff
@@ -205,7 +206,7 @@ def ext_pillar(minion_id,
                        'is not formed as a list'.format(re_minion_id))
                 log.error(msg)
             else:
-                included_data = {}
+                included_data = copy.deepcopy(result)
```
Comment (pcdummy) [inline salt/pillar/mongo.py]: I'm not sure about this line. It works for me (TM), but people who have enabled "pillar_merge_lists" would have duplicated lists (if there are any in the doc!). What do you think about it?

Comment (pcdummy) [inline salt/pillar/mongo.py]: I'm not sure about this line. It works for me (TM), but people who have enabled "pillar_merge_lists" would have duplicated lists (if there are any in the doc!). What do you think about it?

File: salt/pillar/mongo.py [hunk 3]
```diff
@@ -205,7 +206,8 @@ def ext_pillar(minion_id,
...
```
Comment (pcdummy) [inline salt/pillar/mongo.py]: I'm not sure about this line. It works for me (TM), but people who have enabled "pillar_merge_lists" would have duplicated lists (if there are any in the doc!). What do you think about it?

**Comment by pcdummy** [comment]
Have written another howto about my latest patch here: 
https://rene.jochums.at/having-fun-with-saltstack-and-ext_pillar-mongo

**Comment by cachedout** [comment]
@pcdummy We just needed these lint errors fixed: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/3261/violations/file/salt/pillar/mongo.py/

Thanks!

**Comment by thatch45** [comment]
@pcdummy I am hesitant to bring this in, it seems to be a special use case that is very specific. I think that this should be in salt-contrib or a separate repo for now before we merge it into mainline.

**Comment by cachedout** [comment]
I'm in the same camp as @thatch45 here. I'd rather this be maintained outside of salt-core. Bear in mind that we do pass in the compiled pillar into ext pillars, which is how the majority of these use cases are handled: https://docs.saltstack.com/en/latest/topics/development/external_pillars.html#ext-pillar

**Comment by pcdummy** [comment]
@thatch45 @cachedout Fine, will bring it in a extra repo and include over gitfs. :+1:

**Comment by cachedout** [comment]
@pcdummy Thanks for understanding. Much appreciated. :]

**Comment by pcdummy** [comment]
Have to ask again, whats about my first patch @cachedout @thatch45 https://github.com/saltstack/salt/pull/34566/commits/6ad5587575c0048813022a02efd3f71fae324f81

Can this get included into salt-core ?

**Comment by cachedout** [comment]
@pcdummy We would prefer to have it in contrib.

**Comment by pcdummy** [comment]
ok

**Pull request closed without merging**